From 98aa594e410820d36728c328efbbd1c0113744d2 Mon Sep 17 00:00:00 2001
From: Oliver Bruns <obruns@gmail.com>
Date: Tue, 3 Jun 2025 16:45:31 +0200
Subject: [PATCH] add minimal support for BzlMod

The two header files are generated by `src/CMakeLists.txt` and added
here for convenience. This entire thing is not production-ready and just
meant for demonstration purposes.
---
 BUILD.bazel                                   | 34 ++++++++
 MODULE.bazel                                  |  8 ++
 include/exiv2/exiv2lib_export.h               | 43 ++++++++++
 include/exiv2/exv_conf.h                      | 80 +++++++++++++++++++
 src/xmp.cpp                                   |  2 +-
 xmpsdk/include/{XMP.incl_cpp => XMP.hpp}      | 16 ++--
 ...TXMPIterator.incl_cpp => TXMPIterator.hpp} |  0
 .../{TXMPMeta.incl_cpp => TXMPMeta.hpp}       |  0
 .../{TXMPUtils.incl_cpp => TXMPUtils.hpp}     |  0
 ...odeInlines.incl_cpp => UnicodeInlines.hpp} |  0
 xmpsdk/src/XMPCore_Impl.cpp                   |  2 +-
 xmpsdk/src/XMPMeta-GetSet.cpp                 |  2 +-
 xmpsdk/src/XMPMeta-Parse.cpp                  |  2 +-
 xmpsdk/src/XMPMeta-Serialize.cpp              |  2 +-
 xmpsdk/src/XMPMeta.cpp                        |  2 +-
 15 files changed, 179 insertions(+), 14 deletions(-)
 create mode 100644 BUILD.bazel
 create mode 100644 MODULE.bazel
 create mode 100644 include/exiv2/exiv2lib_export.h
 create mode 100644 include/exiv2/exv_conf.h
 rename xmpsdk/include/{XMP.incl_cpp => XMP.hpp} (87%)
 rename xmpsdk/include/client-glue/{TXMPIterator.incl_cpp => TXMPIterator.hpp} (100%)
 rename xmpsdk/include/client-glue/{TXMPMeta.incl_cpp => TXMPMeta.hpp} (100%)
 rename xmpsdk/include/client-glue/{TXMPUtils.incl_cpp => TXMPUtils.hpp} (100%)
 rename xmpsdk/src/{UnicodeInlines.incl_cpp => UnicodeInlines.hpp} (100%)

diff --git a/BUILD.bazel b/BUILD.bazel
new file mode 100644
index 0000000..cc1f27b
--- /dev/null
+++ b/BUILD.bazel
@@ -0,0 +1,34 @@
+cc_library(
+    name = "exiv2",
+    srcs = glob(
+        [
+            "src/*.cpp",
+            "src/*.hpp",
+            "src/*.h",
+            "xmpsdk/src/**/*.cpp",
+            "xmpsdk/src/**/*.hpp",
+            "xmpsdk/src/**/*.h",
+            "xmpsdk/include/**/*.hpp",
+            "xmpsdk/include/**/*.h",
+        ],
+    ),
+    hdrs = glob(["include/exiv2/*.hpp"]) + [
+        "include/exiv2/config.h",
+        "include/exiv2/exiv2lib_export.h",
+        "include/exiv2/exv_conf.h",
+    ],
+    includes = [
+        "include/exiv2",
+        "xmpsdk/include/",
+    ],
+    local_defines = [
+        'EXV_LOCALEDIR=\\"/usr/share/locale\\"',
+    ],
+    #hdrs = ["include/exiv2/exiv2.hpp"],
+    strip_include_prefix = "include",
+    visibility = ["//visibility:public"],
+    deps = [
+        "@libexpat",
+        "@zlib",
+    ],
+)
diff --git a/MODULE.bazel b/MODULE.bazel
new file mode 100644
index 0000000..cf75c3b
--- /dev/null
+++ b/MODULE.bazel
@@ -0,0 +1,8 @@
+module(
+    name = "exiv2",
+    version = "0.28.5",
+    compatibility_level = 1,
+)
+
+bazel_dep(name = "libexpat", version = "2.7.1")
+bazel_dep(name = "zlib", version = "1.3.1.bcr.6")
diff --git a/include/exiv2/exiv2lib_export.h b/include/exiv2/exiv2lib_export.h
new file mode 100644
index 0000000..5eebda5
--- /dev/null
+++ b/include/exiv2/exiv2lib_export.h
@@ -0,0 +1,43 @@
+
+#ifndef EXIV2API_H
+#define EXIV2API_H
+
+#ifdef exiv2lib_STATIC
+#  define EXIV2API
+#  define EXIV2LIB_NO_EXPORT
+#else
+#  ifndef EXIV2API
+#    ifdef exiv2lib_EXPORTS
+        /* We are building this library */
+#      define EXIV2API __attribute__((visibility("default")))
+#    else
+        /* We are using this library */
+#      define EXIV2API __attribute__((visibility("default")))
+#    endif
+#  endif
+
+#  ifndef EXIV2LIB_NO_EXPORT
+#    define EXIV2LIB_NO_EXPORT __attribute__((visibility("hidden")))
+#  endif
+#endif
+
+#ifndef EXIV2LIB_DEPRECATED
+#  define EXIV2LIB_DEPRECATED __attribute__ ((__deprecated__))
+#endif
+
+#ifndef EXIV2LIB_DEPRECATED_EXPORT
+#  define EXIV2LIB_DEPRECATED_EXPORT EXIV2API EXIV2LIB_DEPRECATED
+#endif
+
+#ifndef EXIV2LIB_DEPRECATED_NO_EXPORT
+#  define EXIV2LIB_DEPRECATED_NO_EXPORT EXIV2LIB_NO_EXPORT EXIV2LIB_DEPRECATED
+#endif
+
+/* NOLINTNEXTLINE(readability-avoid-unconditional-preprocessor-if) */
+#if 0 /* DEFINE_NO_DEPRECATED */
+#  ifndef EXIV2LIB_NO_DEPRECATED
+#    define EXIV2LIB_NO_DEPRECATED
+#  endif
+#endif
+
+#endif /* EXIV2API_H */
diff --git a/include/exiv2/exv_conf.h b/include/exiv2/exv_conf.h
new file mode 100644
index 0000000..eb67f3a
--- /dev/null
+++ b/include/exiv2/exv_conf.h
@@ -0,0 +1,80 @@
+// File generated by cmake from cmake/config.h.cmake.
+
+#ifndef _EXV_CONF_H_
+#define _EXV_CONF_H_
+
+// Define to 1 if you want to use libcurl in httpIO.
+/* #undef EXV_USE_CURL */
+
+// Define to 1 if you want to enable filesystem access
+#define EXV_ENABLE_FILESYSTEM
+
+// Define if you require webready support.
+/* #undef EXV_ENABLE_WEBREADY */
+
+// Define if you want translation of program messages to the user's native language
+// #define EXV_ENABLE_NLS
+
+// Define if you want to enable the decoding of video metadata
+// #define EXV_ENABLE_VIDEO
+
+// Define if you want BMFF support.
+// #define EXV_ENABLE_BMFF
+
+// Define if you want to use the inih library.
+// #define EXV_ENABLE_INIH
+
+// Define if you have the strerror_r function.
+#define EXV_HAVE_STRERROR_R
+
+// Define if the strerror_r function returns char*.
+#define EXV_STRERROR_R_CHAR_P
+
+/* Define to `const' or to empty, depending on the second argument of `iconv'. */
+/* #undef ICONV_ACCEPTS_CONST_INPUT */
+
+#if defined(__NetBSD__)
+#include <sys/param.h>
+#if __NetBSD_Prereq__(9,99,17)
+#define NETBSD_POSIX_ICONV 1
+#else
+#define NETBSD_POSIX_ICONV 0
+#endif
+#endif
+
+#if defined(ICONV_ACCEPTS_CONST_INPUT) || (defined(__NetBSD__) && !NETBSD_POSIX_ICONV)
+#define EXV_ICONV_CONST const
+#else
+#define EXV_ICONV_CONST
+#endif
+
+// Define if you have the zlib library.
+#define EXV_HAVE_LIBZ
+
+// Define if you have the brotli library.
+/* #undef EXV_HAVE_BROTLI */
+
+/* Define if you have (Exiv2/xmpsdk) Adobe XMP Toolkit. */
+#define EXV_HAVE_XMP_TOOLKIT
+
+/* Define to the full name of this package. */
+#define EXV_PACKAGE_NAME "exiv2"
+
+/* Define to the full name and version of this package. */
+#define EXV_PACKAGE_STRING "exiv2 0.28.5"
+
+/* Define to the version of this package. */
+#define EXV_PACKAGE_VERSION "0.28.5"
+
+#define EXIV2_MAJOR_VERSION (0U)
+#define EXIV2_MINOR_VERSION (28U)
+#define EXIV2_PATCH_VERSION (5U)
+#define EXIV2_TWEAK_VERSION (U)
+
+// Definition to enable translation of Nikon lens names.
+#define EXV_HAVE_LENSDATA
+
+// Define if you have the iconv function.
+#define EXV_HAVE_ICONV
+
+#endif /* !_EXV_CONF_H_ */
diff --git a/src/xmp.cpp b/src/xmp.cpp
index 4a7f431..34f5f54 100644
--- a/src/xmp.cpp
+++ b/src/xmp.cpp
@@ -21,7 +21,7 @@
 #else
 #include <XMPSDK.hpp>
 #endif
-#include <XMP.incl_cpp>
+#include <XMP.hpp>
 #endif  // EXV_HAVE_XMP_TOOLKIT
 
 #ifdef EXV_HAVE_XMP_TOOLKIT
diff --git a/xmpsdk/include/XMP.incl_cpp b/xmpsdk/include/XMP.hpp
similarity index 87%
rename from xmpsdk/include/XMP.incl_cpp
rename to xmpsdk/include/XMP.hpp
index 6501be9..899cc31 100644
--- a/xmpsdk/include/XMP.incl_cpp
+++ b/xmpsdk/include/XMP.hpp
@@ -1,5 +1,5 @@
-#ifndef __XMP_incl_cpp__
-#define __XMP_incl_cpp__ 1
+#ifndef __XMP_inc_hpp__
+#define __XMP_inc_hpp__ 1
 
 // =================================================================================================
 // ADOBE SYSTEMS INCORPORATED
@@ -11,7 +11,7 @@
 // =================================================================================================
 
 //  ================================================================================================
-/// \file XMP.incl_cpp
+/// \file XMP.hpp
 /// \brief Overall client glue file for the XMP toolkit.
 ///
 /// This is an overall client source file of XMP toolkit glue, the only XMP-specific one that
@@ -24,7 +24,7 @@
 ///
 /// \code
 ///  #include <string>
-///  #include "XMP.incl_cpp"
+///  #include "XMP.hpp"
 /// \endcode
 
 
@@ -47,9 +47,9 @@
 
     // We're using a single out of line instantiation. Do it here.
 
-    #include "client-glue/TXMPMeta.incl_cpp"
-    #include "client-glue/TXMPIterator.incl_cpp"
-    #include "client-glue/TXMPUtils.incl_cpp"
+    #include "client-glue/TXMPMeta.hpp"
+    #include "client-glue/TXMPIterator.hpp"
+    #include "client-glue/TXMPUtils.hpp"
     template class TXMPMeta <TXMP_STRING_TYPE>;
     template class TXMPIterator <TXMP_STRING_TYPE>;
     template class TXMPUtils <TXMP_STRING_TYPE>;
@@ -60,4 +60,4 @@
     #pragma warning ( pop )
 #endif
 
-#endif // __XMP_incl_cpp__
+#endif // __XMP_inc_hpp__
diff --git a/xmpsdk/include/client-glue/TXMPIterator.incl_cpp b/xmpsdk/include/client-glue/TXMPIterator.hpp
similarity index 100%
rename from xmpsdk/include/client-glue/TXMPIterator.incl_cpp
rename to xmpsdk/include/client-glue/TXMPIterator.hpp
diff --git a/xmpsdk/include/client-glue/TXMPMeta.incl_cpp b/xmpsdk/include/client-glue/TXMPMeta.hpp
similarity index 100%
rename from xmpsdk/include/client-glue/TXMPMeta.incl_cpp
rename to xmpsdk/include/client-glue/TXMPMeta.hpp
diff --git a/xmpsdk/include/client-glue/TXMPUtils.incl_cpp b/xmpsdk/include/client-glue/TXMPUtils.hpp
similarity index 100%
rename from xmpsdk/include/client-glue/TXMPUtils.incl_cpp
rename to xmpsdk/include/client-glue/TXMPUtils.hpp
diff --git a/xmpsdk/src/UnicodeInlines.incl_cpp b/xmpsdk/src/UnicodeInlines.hpp
similarity index 100%
rename from xmpsdk/src/UnicodeInlines.incl_cpp
rename to xmpsdk/src/UnicodeInlines.hpp
diff --git a/xmpsdk/src/XMPCore_Impl.cpp b/xmpsdk/src/XMPCore_Impl.cpp
index b810bfe..3e10aea 100644
--- a/xmpsdk/src/XMPCore_Impl.cpp
+++ b/xmpsdk/src/XMPCore_Impl.cpp
@@ -11,7 +11,7 @@
 #include "XMPCore_Impl.hpp"
 #include "XMPMeta.hpp"	// *** For use of GetNamespacePrefix in FindSchemaNode.
 
-#include "UnicodeInlines.incl_cpp"
+#include "UnicodeInlines.hpp"
 
 #include <algorithm>
 
diff --git a/xmpsdk/src/XMPMeta-GetSet.cpp b/xmpsdk/src/XMPMeta-GetSet.cpp
index 9972f9e..b232bbf 100644
--- a/xmpsdk/src/XMPMeta-GetSet.cpp
+++ b/xmpsdk/src/XMPMeta-GetSet.cpp
@@ -17,7 +17,7 @@
 #include "XMPUtils.hpp"
 
 #include "XMP_Version.h"
-#include "UnicodeInlines.incl_cpp"
+#include "UnicodeInlines.hpp"
 #include "UnicodeConversions.hpp"
 #include "ExpatAdapter.hpp"
 
diff --git a/xmpsdk/src/XMPMeta-Parse.cpp b/xmpsdk/src/XMPMeta-Parse.cpp
index 6959693..22aa63b 100644
--- a/xmpsdk/src/XMPMeta-Parse.cpp
+++ b/xmpsdk/src/XMPMeta-Parse.cpp
@@ -15,7 +15,7 @@
 #include "XMPMeta.hpp"
 #include "XMPUtils.hpp"
 
-#include "UnicodeInlines.incl_cpp"
+#include "UnicodeInlines.hpp"
 #include "UnicodeConversions.hpp"
 #include "ExpatAdapter.hpp"
 
diff --git a/xmpsdk/src/XMPMeta-Serialize.cpp b/xmpsdk/src/XMPMeta-Serialize.cpp
index 70a7b02..39233e5 100644
--- a/xmpsdk/src/XMPMeta-Serialize.cpp
+++ b/xmpsdk/src/XMPMeta-Serialize.cpp
@@ -14,7 +14,7 @@
 #include "XMPMeta.hpp"
 
 #include "XMP_Version.h"
-#include "UnicodeInlines.incl_cpp"
+#include "UnicodeInlines.hpp"
 #include "UnicodeConversions.hpp"
 
 #if XMP_DebugBuild
diff --git a/xmpsdk/src/XMPMeta.cpp b/xmpsdk/src/XMPMeta.cpp
index 30ff4a5..aabb42b 100644
--- a/xmpsdk/src/XMPMeta.cpp
+++ b/xmpsdk/src/XMPMeta.cpp
@@ -18,7 +18,7 @@
 #include "XMPIterator.hpp"
 #include "XMPUtils.hpp"
 #include "XMP_Version.h"
-#include "UnicodeInlines.incl_cpp"
+#include "UnicodeInlines.hpp"
 #include "UnicodeConversions.hpp"
 
 #include <algorithm>	// For sort and stable_sort.
-- 
2.49.0

